### DISCUSSION TITLE
Test parallelization using Config.connect_to

### DISCUSSION THREAD

**Discussion by asgerb** [mongodb/mongoid#5246] [Ideas]
I've implemented test parallelization in my app using this module to "fork" the database, following the example found in `ActiveRecord`:

```ruby
require "active_support/testing/parallelization"
module Mongoid
  module TestDatabases # :nodoc:
    ActiveSupport::Testing::Parallelization.after_fork_hook do |i|
      Mongoid::Config.connect_to("#{database_name}-#{i}")
    end

    def self.database_name
      Mongoid::Config.clients["default"]["database"]
    end
  end
end
```

This worked fine so far, but I've been running into issues lately as I'm transitioning to running tests in containers. 
Specifically the test suite would try to connect to `localhost:27017` no matter what values I would configure in `mongoid.yml`. After some debugging I found out that issue stems from my use of [`Mongoid::Config.connect_to`](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/config.rb#L136), which changes the hosts to `localhost:27017`.

It would be great to make it either possible to pass in the hosts, or possibly to read the hosts from the config if possible? We could then fall back to `localhost:27017` if that fails. Would be open to make a PR following any guidance on which path to take. 

For anyone stumbling on this looking for a solution, here's my dirty hack to make it work:

```ruby
require "active_support/testing/parallelization"
module Mongoid
  module TestDatabases # :nodoc:
    ActiveSupport::Testing::Parallelization.after_fork_hook do |i|
      Mongoid::Config.send("clients=", {
        default: {
          database: "#{database_name}-#{i}",
          hosts: database_hosts,
          options: { read: { mode: :primary }}
        }
      })
    end

    def self.database_name
      Mongoid::Config.clients["default"]["database"]
    end

    def self.database_hosts
      Mongoid::Config.clients["default"]["hosts"]
    end
  end
end
```

**Comment by p-mongo** [comment]
Hi @asgerb , 

Programmatic configuration of Mongoid is described in https://www.mongodb.com/docs/mongoid/master/reference/configuration/#loading-mongoid-configuration (`Mongoid.configure` etc.). Does this answer your question?

**Comment by donnykurnia** [comment]
Thanks @asgerb 

I put the code in `test_helper.rb` inside `parallelize_setup` block:

```
class ActiveSupport::TestCase
  ...

  # Run tests in parallel with specified workers
  parallelize_setup do |worker|
    # setup databases
    c = Mongoid.load!('config/mongoid.yml')
    if c['clients']['default']['uri'].present?
      uri = URI(c['clients']['default']['uri'])
      uri.path += "-#{worker}"
      Mongoid::Config.send("clients=", {
        default: {
          uri: uri.to_s
        }
      })
    else
      database = c['clients']['default']['database']
      hosts = c['clients']['default']['hosts']
      Mongoid::Config.send("clients=", {
        default: {
          database: "#{database}-#{worker}",
          hosts: hosts
        }
      })
    end
  end

  # parallelize_teardown do |worker|
  #   # cleanup databases
  # end
  parallelize(workers: :number_of_processors)
  ...
end
```

I use the `uri` to connect to MongoDB in docker.
